src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKey.cs

```diff
@@ -0,0 +1,120 @@
+        private static readonly LocalizableString s_localizableMessage =
+            new LocalizableResourceString(nameof(Resx.DoNotGuardDictionaryRemoveByContainsKeyMessage), Resx.ResourceManager, typeof(Resx));
+        private static readonly LocalizableString s_localizableDescription =
+            new LocalizableResourceString(nameof(Resx.DoNotGuardDictionaryRemoveByContainsKeyDescription), Resx.ResourceManager, typeof(Resx));
+
+        public const string AdditionalDocumentLocationInfoSeparator = ";;";
+
+        public const string ConditionalOperation = nameof(ConditionalOperation);
+        public const string ChildStatementOperation = nameof(ChildStatementOperation);
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false,
+            additionalCustomTags: WellKnownDiagnosticTags.Unnecessary);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericDictionary2, out var dictionaryType))
+                return;
+
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.Invocation);
+
+            static void AnalyzeOperation(OperationAnalysisContext context)
+            {
+                var invocationOperation = (IInvocationOperation)context.Operation;
+
+                if (invocationOperation.TargetMethod.Name != "ContainsKey")
+                    return;
+
+                if (invocationOperation.Parent is not IConditionalOperation parentConditionalOperation)
+                    return;
+
+                // we only want to report this diagnostic if the Contains/Remove pair is all there is
+
+                if (parentConditionalOperation.WhenFalse == null &&
+                    parentConditionalOperation.WhenTrue.Children.HasExactly(1))
+                {
+                    var properties = ImmutableDictionary.CreateBuilder<string, string?>();
+                    properties[ConditionalOperation] = CreateLocationInfo(parentConditionalOperation.Syntax);
```

Comment:
Not sure if marking the location manually would cause any issue, as far as i know, we get the location or the operation from the code fixer context, i am pretty sure you can get this info from there easily, but might not that easy to get the 2nd location (for Remove), but i think you can resolve it even without the location or `Remove`

Reply:
With that being said I am not completely opposing this solution, it just might be not the suggested way for analyzers, tagging @mavasani for input

Reply:
>might not that easy to get the 2nd location (for Remove)

Right. I was trying to store two locations, and `DiagnosticDescriptorCreationAnalyzer` seemed to approach that problem this way. (https://github.com/dotnet/roslyn-analyzers/blob/5102a0d89ef178b0ca741634ee006597bf5df2ed/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs#L545)

I _could_ just store the condition and then find the invocation a second time.

>i think you can resolve it even without the location or Remove. 

I'm not sure how I would approach it without _any_ location?

Reply:
> I'm not sure how I would approach it without any location?

Sorry i didn't mean without any location, i meant without saving/passing any location info, especially for `Remove`, because you can populate the `Remove` operation from `ContainsKey`
```cs
    public MyClass()
    {
        if ([|MyDictionary.ContainsKey(""Key"")|])
            MyDictionary.Remove(""Key"");
    }
```
As you know you can get the location of `MyDictionary.ContainsKey` from the context, then you can replace `ContainsKey` with `Remove` and use it for the fixer, but it can work only for the simplest case. 

Anyway, I am OK with the existing solution